### PR TITLE
Promote GH gauge constraint from test to function

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

Move GH gauge constraint from a test file (VerifyGrSolution.hpp) to GeneralizedHarmonic/Constraints.?pp, and test by checking for Kerr-Schild and with fuzzy testing.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 1]
```diff
@@ -26,6 +41,101 @@ void test_three_index_constraint(const DataType& used_for_size) noexcept {
...
+  const auto& d_shift =
+      get<typename Solution::template DerivShift<DataVector>>(vars);
+  const auto& dt_shift = get<Tags::dt<gr::Tags::Shift<3>>>(vars);
+  const auto& spatial_metric = get<gr::Tags::SpatialMetric<3>>(vars);
+  const auto& dt_spatial_metric =
+      get<Tags::dt<gr::Tags::SpatialMetric<3>>>(vars);
+  const auto& d_spatial_metric =
+      get<typename Solution::template DerivSpatialMetric<DataVector>>(vars);
+
+  // Check those quantities that should vanish identically.
```
Comment (geoffrey4444) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: Originally, Mark's code checked that the solution was static. That's not necessary to test the gauge constraint, right? But these checks will pass, because I'm only checking Kerr Schild.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp [hunk 2]
```diff
@@ -37,4 +37,44 @@ void three_index_constraint(
...
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void gauge_constraint(
```
Comment (osheamonn) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: Is this version of the function tested?

Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: Unless I screwed something up. The check_with_random_values test calls this one (return by value), which then calls the other one (return by pointer), thus testing both. Or at a least, that's what's supposed to happen...did I screw up somewhere?

Comment (osheamonn) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: I'd like to see this function explicitly tested though, since the fact that one calls the other is an implementation detail. Since you've two tests, the analytic and the random version, I'd be fine with just adding a call to this version into the analytic test, and checking that it also gives zero.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp [hunk 3]
```diff
@@ -37,4 +37,44 @@ void three_index_constraint(
...
+ * [Eq. (40) of http://arXiv.org/abs/gr-qc/0512093v3],
+ * \f[
+ * C_a = H_a + g^{ij} \Phi_{ija} + t^b \Pi_{ba}
+ * - \frac{1}{2} g^i_a \psi^{bc} \Phi_{ibc}
+ * - \frac{1}{2} t_a \psi^{bc} \Pi_{bc},
+ * \f]
+ * where \f$H_a\f$ is the gauge function,
+ * \f$\psi_{ab}\f$ is the spacetime metric,
+ * \f$\Pi_{ab}=-t^c\partial_c \psi_{ab}\f$, and
+ * \f$\Phi_{iab} = \partial_i\psi_{ab}\f$; \f$t^a\f$ is the timelike unit
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: Why does the one-form have an up index?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 4]
```diff
@@ -61,3 +171,49 @@ SPECTRE_TEST_CASE(
...
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.GeneralizedHarmonic.GaugeConstraint",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/GeneralizedHarmonic/"};
+
+  // Test the gauge constraint against Kerr Schild
+  const double mass = 1.4;
+  const std::array<double, 3> spin{{0.4, 0.3, 0.2}};
+  const std::array<double, 3> center{{0.2, 0.3, 0.4}};
+  gr::Solutions::KerrSchild solution(mass, spin, center);
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: the solution can't be const?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 5]
```diff
@@ -26,6 +41,101 @@ void test_three_index_constraint(const DataType& used_for_size) noexcept {
...
+      make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  const double t = 4.444;  // Arbitrary time for time-independent solution.
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: can you set this to `0.0` please?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp [hunk 6]
```diff
@@ -26,6 +41,101 @@ void test_three_index_constraint(const DataType& used_for_size) noexcept {
...
+  const auto& phi = GeneralizedHarmonic::phi(lapse, d_lapse, shift, d_shift,
+                                             spatial_metric, d_spatial_metric);
+  const auto& pi = GeneralizedHarmonic::pi(
+      lapse, dt_lapse, shift, dt_shift, spatial_metric, dt_spatial_metric, phi);
+  const auto normal_one_form =
+      gr::spacetime_normal_one_form<3, Frame::Inertial>(lapse);
+  const auto& normal_vector = gr::spacetime_normal_vector(lapse, shift);
+
+  // Get the constraint, and check that it vanishes
+  auto constraint = GeneralizedHarmonic::gauge_constraint(
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: can't this be `const`?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 7]
```diff
@@ -37,6 +37,65 @@ void three_index_constraint(
...
+void gauge_constraint(
+    gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  *constraint = make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Why do you need to do this if the allocation is already set? Doing this allocation completely defeats the purpose of the return-by-reference approach. You should check if the allocation is necessary and only do it if it is.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 8]
```diff
@@ -37,6 +37,65 @@ void three_index_constraint(
...
+  gauge_constraint<SpatialDim, Frame, DataType>(
+      &constraint, gauge_function, spacetime_normal_one_form,
+      spacetime_normal_vector, inverse_spatial_metric,
+      inverse_spacetime_metric, pi, phi);
+  return constraint;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void gauge_constraint(
+    gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Can you make the pointer `const`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp (hunk 5 above)
Comment (geoffrey4444) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: I can, but isn’t a nonzero choice better in general?

Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: I don't see how it's better if it won't matter. `signalingNaN` would be better, though.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp (hunk 3 above)
Comment (hrueter) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: @nilsdeppe I guess he means a vector field and not a 1-form.
It think it is sensible to use the vector field here.

Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: Typo; I meant to say vector

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp [hunk 9]
```diff
@@ -53,7 +53,7 @@ void three_index_constraint(
  * \f$\psi_{ab}\f$ is the spacetime metric,
  * \f$\Pi_{ab}=-t^c\partial_c \psi_{ab}\f$, and
  * \f$\Phi_{iab} = \partial_i\psi_{ab}\f$; \f$t^a\f$ is the timelike unit
- * normal one form to the spatial slice, \f$g^{ij}\f$ is the inverse spatial
+ * normal vector form to the spatial slice, \f$g^{ij}\f$ is the inverse spatial
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: maybe `vector` instead of `vector form`?

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash when you fix my last small comment :)

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@wthrowe @kidder @osheamonn @markscheel could you please look at this whenever you get time? Thanks! :)

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 10]
```diff
@@ -37,6 +37,63 @@ void three_index_constraint(
...
+      for (size_t j = 0; j < SpatialDim; ++j) {
+        (*constraint).get(a) +=
+            inverse_spatial_metric.get(i, j) * phi.get(i, j + 1, a);
+      }
+    }
+    for (size_t b = 0; b < SpatialDim + 1; ++b) {
+      (*constraint).get(a) += spacetime_normal_vector.get(b) * pi.get(b, a);
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
+        (*constraint).get(a) -= 0.5 * spacetime_normal_one_form.get(a) *
+                                   pi.get(b, c) *
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: can this save operations by computing $\pi_{bc} g^{bc}$ first in a loop, then subtracting from the constraint outside the double-loop? (if so, same on line 90)

**Review by moxcodes — COMMENTED** [review]
Just a quick question - if it's not worth it at this stage, no worries.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp (hunk 10 above)
Comment (wthrowe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: @nilsdeppe Thoughts?

**Review by wthrowe — APPROVED** [review]
I'm fine with this, but I'll let Nils take a look at the last comment and see if if he wants a change.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: I think this would add an allocation, so it's not obvious to me it'll be faster and would need benchmarking. I'd be fine with @geoffrey4444 filing an issue and coming back to it later :)

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 11]
```diff
@@ -37,6 +37,63 @@ void three_index_constraint(
...
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  for (size_t a = 0; a < SpatialDim + 1; ++a) {
+    (*constraint).get(a) = gauge_function.get(a);
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:70]: `constraint->get(a)` and many times below (arrow operator serves to dereference and access).

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 12]
```diff
@@ -37,6 +37,63 @@ void three_index_constraint(
...
+            inverse_spatial_metric.get(i, j) * phi.get(i, j + 1, a);
+      }
+    }
+    for (size_t b = 0; b < SpatialDim + 1; ++b) {
+      (*constraint).get(a) += spacetime_normal_vector.get(b) * pi.get(b, a);
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
+        (*constraint).get(a) -= 0.5 * spacetime_normal_one_form.get(a) *
+                                   pi.get(b, c) *
+                                   inverse_spacetime_metric.get(b, c);
+        if (a > 0) {
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:83]: can avoid `if` statement by doing this work in a separate loop `for(size_t a=1;...)`

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp [hunk 13]
```diff
@@ -37,6 +37,63 @@ void three_index_constraint(
...
+    (*constraint).get(a) = gauge_function.get(a);
+    for (size_t i = 0; i < SpatialDim; ++i) {
+      for (size_t j = 0; j < SpatialDim; ++j) {
+        (*constraint).get(a) +=
+            inverse_spatial_metric.get(i, j) * phi.get(i, j + 1, a);
+      }
+    }
+    for (size_t b = 0; b < SpatialDim + 1; ++b) {
+      (*constraint).get(a) += spacetime_normal_vector.get(b) * pi.get(b, a);
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
```
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:79]: because all expressions summed over are symmetric in b and c, you can start c at b+1 (with appropriate factors of 2 in each expression) and add the diagonal term afterward for savings.

**Review by moxcodes — COMMENTED** [review]
Some thoughts for possible optimization if/when desired. (code review with @markscheel @fmahebert @mgiesler)

**Comment by geoffrey4444** [comment]
Should I create an issue to look into the suggested optimizations? Or would you like me to do any of these before we push the merge button? I'm fine either way.

**Comment by kidder** [comment]
you can create an issue
